File: sdks/sandbox/csharp/src/OpenSandbox/SandboxManager.cs
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using OpenSandbox.Config;
using OpenSandbox.Factory;
using OpenSandbox.Models;
using OpenSandbox.Services;
using OpenSandbox.Core;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace OpenSandbox;

/// <summary>
/// Administrative interface for managing sandboxes.
/// </summary>
/// <remarks>
/// This type is intended for administrative lifecycle operations (list, inspect, pause, resume, kill, renew).
/// Dispose the manager when finished to release local SDK resources.
/// </remarks>
public sealed class SandboxManager : IAsyncDisposable
{
    private readonly ISandboxes _sandboxes;
    private readonly ConnectionConfig _connectionConfig;
    private readonly HttpClientProvider _httpClientProvider;
    private readonly ILogger _logger;
    private bool _disposed;

    private SandboxManager(
        ISandboxes sandboxes,
        ConnectionConfig connectionConfig,
        HttpClientProvider httpClientProvider,
        ILoggerFactory loggerFactory)
    {
        _sandboxes = sandboxes;
        _connectionConfig = connectionConfig;
        _httpClientProvider = httpClientProvider;
        _logger = (loggerFactory ?? NullLoggerFactory.Instance).CreateLogger("OpenSandbox.SandboxManager");
    }

    /// <summary>
    /// Creates a new sandbox manager.
    /// </summary>
    /// <param name="options">Optional configuration options.</param>
    /// <returns>A new sandbox manager instance.</returns>
    /// <exception cref="SandboxException">Thrown when manager initialization fails.</exception>
    public static SandboxManager Create(SandboxManagerOptions? options = null)
    {
        var connectionConfig = options?.ConnectionConfig ?? new ConnectionConfig();
        var lifecycleBaseUrl = connectionConfig.GetBaseUrl();
        var adapterFactory = options?.AdapterFactory ?? DefaultAdapterFactory.Create();
        var loggerFactory = options?.Diagnostics?.LoggerFactory ?? NullLoggerFactory.Instance;
        var httpClientProvider = new HttpClientProvider(connectionConfig, loggerFactory);
        var logger = loggerFactory.CreateLogger("OpenSandbox.SandboxManager");
        logger.LogInformation("Creating sandbox manager");

        try
        {
            var lifecycleStack = adapterFactory.CreateLifecycleStack(new CreateLifecycleStackOptions
            {
                ConnectionConfig = connectionConfig,
                LifecycleBaseUrl = lifecycleBaseUrl,
                HttpClientProvider = httpClientProvider,
                LoggerFactory = loggerFactory
            });

            return new SandboxManager(lifecycleStack.Sandboxes, connectionConfig, httpClientProvider, loggerFactory);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to create sandbox manager");
            httpClientProvider.Dispose();
            throw;
        }
    }

    /// <summary>
    /// Lists sandboxes with optional filtering.
    /// </summary>
    /// <param name="filter">Optional filter criteria.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The list of sandboxes.</returns>
    /// <exception cref="SandboxApiException">Thrown when the sandbox API returns an error.</exception>
    public Task<ListSandboxesResponse> ListSandboxInfosAsync(
        SandboxFilter? filter = null,
        CancellationToken cancellationToken = default)
    {
        return _sandboxes.ListSandboxesAsync(new ListSandboxesParams
        {
            States = filter?.States,
            Metadata = filter?.Metadata,
            Page = filter?.Page,
            PageSize = filter?.PageSize
        }, cancellationToken);
    }

    /// <summary>
    /// Gets information about a specific sandbox.
    /// </summary>
    /// <param name="sandboxId">The sandbox ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The sandbox information.</returns>
    /// <exception cref="InvalidArgumentException">Thrown when <paramref name="sandboxId"/> is null or empty.</exception>
    /// <exception cref="SandboxApiException">Thrown when the sandbox API returns an error.</exception>
    public Task<SandboxInfo> GetSandboxInfoAsync(
        string sandboxId,
        CancellationToken cancellationToken = default)
    {
        _logger.LogDebug("Fetching sandbox info: {SandboxId}", sandboxId);
        return _sandboxes.GetSandboxAsync(sandboxId, cancellationToken);
    }

    /// <summary>
    /// Terminates a sandbox.
    /// </summary>
    /// <param name="sandboxId">The sandbox ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <exception cref="InvalidArgumentException">Thrown when <paramref name="sandboxId"/> is null or empty.</exception>
    /// <exception cref="SandboxApiException">Thrown when the sandbox API returns an error.</exception>
    public Task KillSandboxAsync(
        string sandboxId,
        CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Killing sandbox: {SandboxId}", sandboxId);
        return _sandboxes.DeleteSandboxAsync(sandboxId, cancellationToken);
    }

    /// <summary>
    /// Pauses a sandbox.
    /// </summary>
    /// <param name="sandboxId">The sandbox ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <exception cref="InvalidArgumentException">Thrown when <paramref name="sandboxId"/> is null or empty.</exception>
    /// <exception cref="SandboxApiException">Thrown when the sandbox API returns an error.</exception>
    public Task PauseSandboxAsync(
        string sandboxId,
        CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Pausing sandbox: {SandboxId}", sandboxId);
        return _sandboxes.PauseSandboxAsync(sandboxId, cancellationToken);
    }

    /// <summary>
    /// Resumes a paused sandbox.
    /// </summary>
    /// <param name="sandboxId">The sandbox ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <exception cref="InvalidArgumentException">Thrown when <paramref name="sandboxId"/> is null or empty.</exception>
    /// <exception cref="SandboxApiException">Thrown when the sandbox API returns an error.</exception>
    public Task ResumeSandboxAsync(
        string sandboxId,
        CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Resuming sandbox: {SandboxId}", sandboxId);
        return _sandboxes.ResumeSandboxAsync(sandboxId, cancellationToken);
    }

    /// <summary>
    /// Renews the expiration time of a sandbox.
    /// </summary>
    /// <param name="sandboxId">The sandbox ID.</param>
    /// <param name="timeoutSeconds">The new timeout in seconds from now.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <exception cref="InvalidArgumentException">Thrown when arguments are invalid.</exception>
    /// <exception cref="SandboxApiException">Thrown when the sandbox API returns an error.</exception>
    public async Task RenewSandboxAsync(
        string sandboxId,
        int timeoutSeconds,
        CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Renewing sandbox expiration: {SandboxId} (timeoutSeconds={TimeoutSeconds})", sandboxId, timeoutSeconds);
        var expiresAt = DateTime.UtcNow.AddSeconds(timeoutSeconds).ToString("O");
        await _sandboxes.RenewSandboxExpirationAsync(sandboxId, new RenewSandboxExpirationRequest
        {
            ExpiresAt = expiresAt
        }, cancellationToken).ConfigureAwait(false);
    }

    public Task<SnapshotInfo> CreateSnapshotAsync(
        string sandboxId,
        string? name = null,
        CancellationToken cancellationToken = default)
    {
        return _sandboxes.CreateSnapshotAsync(
            sandboxId,
            new CreateSnapshotRequest { Name = name },
            cancellationToken);
    }

    public Task<SnapshotInfo> GetSnapshotAsync(
        string snapshotId,
        CancellationToken cancellationToken = default)
    {
        return _sandboxes.GetSnapshotAsync(snapshotId, cancellationToken);
    }

    public Task<ListSnapshotsResponse> ListSnapshotsAsync(
        ListSnapshotsParams? filter = null,
        CancellationToken cancellationToken = default)
    {
        return _sandboxes.ListSnapshotsAsync(filter, cancellationToken);
    }

    public Task DeleteSnapshotAsync(
        string snapshotId,
        CancellationToken cancellationToken = default)
    {
        return _sandboxes.DeleteSnapshotAsync(snapshotId, cancellationToken);
    }

    /// <summary>
    /// Releases resources used by this manager.
    /// </summary>
    public ValueTask DisposeAsync()
    {
        if (_disposed)
        {
            return default;
        }

        _disposed = true;
        _logger.LogDebug("Disposing sandbox manager resources");
        _httpClientProvider.Dispose();
        return default;
    }
}

File: sdks/sandbox/kotlin/sandbox/src/main/kotlin/com/alibaba/opensandbox/sandbox/SandboxManager.kt
/*
 * Copyright 2025 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.opensandbox.sandbox

import com.alibaba.opensandbox.sandbox.config.ConnectionConfig
import com.alibaba.opensandbox.sandbox.domain.exceptions.InvalidArgumentException
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxException
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.PagedSandboxInfos
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.PagedSnapshotInfos
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxFilter
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxInfo
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxRenewResponse
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SnapshotFilter
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SnapshotInfo
import com.alibaba.opensandbox.sandbox.domain.services.Sandboxes
import com.alibaba.opensandbox.sandbox.infrastructure.factory.AdapterFactory
import org.slf4j.LoggerFactory
import java.time.Duration
import java.time.OffsetDateTime

/**
 * Sandbox management interface for administrative operations and monitoring sandbox instances.
 *
 * This class provides a centralized interface for managing sandbox instances,
 * enabling administrative operations and sandbox discovery.
 * It focuses on high-level management operations rather than individual sandbox interactions.
 *
 * ## Key Features
 *
 * - **Sandbox Discovery**: List and filter sandbox instances by various criteria
 * - **Administrative Operations**: Individual sandbox management operations
 * - **Connection Pool Management**: Efficient HTTP client reuse for multiple operations
 *
 * ## Usage Example
 *
 * ```kotlin
 * val manager = SandboxManager.builder()
 *     .connectionConfig(connectionConfig)
 *     .build()
 *
 * // List all running sandboxes
 * val runningSandboxes = manager.listSandboxInfos(
 *     SandboxFilter.builder().state("RUNNING").build()
 * )
 *
 * // Individual operations
 * val sandboxId = "sandbox-id"
 * manager.getSandboxInfo(sandboxId)
 * manager.pauseSandbox(sandboxId)
 * manager.resumeSandbox(sandboxId)
 * manager.killSandbox(sandboxId)
 *
 * // Cleanup
 * manager.close()
 * ```
 *
 * **Note**: This class is designed for administrative operations.
 * For individual sandbox interactions, use the [Sandbox] class directly.
 */
class SandboxManager internal constructor(
    private val sandboxService: Sandboxes,
    private val httpClientProvider: HttpClientProvider,
) : AutoCloseable {
    private val logger = LoggerFactory.getLogger(SandboxManager::class.java)

    /**
     * Provides access to shared httpclient provider
     *
     * Allows retrieving underlying http client resources initialized with connection config
     */
    fun httpClientProvider() = httpClientProvider

    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()

        internal fun create(connectionConfig: ConnectionConfig): SandboxManager {
            val httpClientProvider = HttpClientProvider(connectionConfig)
            val factory = AdapterFactory(httpClientProvider)
            val sandboxService = factory.createSandboxes()
            return SandboxManager(sandboxService, httpClientProvider)
        }
    }

    fun listSandboxInfos(filter: SandboxFilter): PagedSandboxInfos {
        return sandboxService.listSandboxes(filter)
    }

    /**
     * Gets information for a single sandbox by its ID.
     *
     * @param sandboxId Sandbox ID to retrieve information for
     * @return SandboxInfo for the specified sandbox
     * @throws SandboxException if the operation fails
     */
    fun getSandboxInfo(sandboxId: String): SandboxInfo {
        logger.debug("Getting info for sandbox: {}", sandboxId)
        return sandboxService.getSandboxInfo(sandboxId)
    }

    /**
     * Terminates a single sandbox.
     *
     * @param sandboxId Sandbox ID to terminate
     * @throws SandboxException if the operation fails
     */
    fun killSandbox(sandboxId: String) {
        logger.info("Terminating sandbox: {}", sandboxId)
        sandboxService.killSandbox(sandboxId)
        logger.info("Successfully terminated sandbox: {}", sandboxId)
    }

    /**
     * Renew expiration time for a single sandbox.
     *
     * The new expiration time will be set to the current time plus the provided duration.
     *
     * @param sandboxId Sandbox ID to renew
     * @param timeout Duration to add to the current time to set the new expiration
     * @throws SandboxException if the operation fails
     */
    fun renewSandbox(
        sandboxId: String,
        timeout: Duration,
    ): SandboxRenewResponse {
        logger.info("Renew expiration for sandbox {} to {}", sandboxId, OffsetDateTime.now().plus(timeout))
        return sandboxService.renewSandboxExpiration(sandboxId, OffsetDateTime.now().plus(timeout))
    }

    /**
     * Pauses a single sandbox while preserving its state.
     *
     * @param sandboxId Sandbox ID to pause
     * @throws SandboxException if the operation fails
     */
    fun pauseSandbox(sandboxId: String) {
        logger.info("Pausing sandbox: {}", sandboxId)
        sandboxService.pauseSandbox(sandboxId)
    }

    /**
     * Resumes a previously paused sandbox.
     *
     * @param sandboxId Sandbox ID to resume
     * @throws SandboxException if the operation fails
     */
    fun resumeSandbox(sandboxId: String) {
        logger.info("Resuming sandbox: {}", sandboxId)
        sandboxService.resumeSandbox(sandboxId)
    }

    fun createSnapshot(
        sandboxId: String,
        name: String? = null,
    ): SnapshotInfo = sandboxService.createSnapshot(sandboxId, name)

    fun getSnapshot(snapshotId: String): SnapshotInfo = sandboxService.getSnapshot(snapshotId)

    fun listSnapshots(filter: SnapshotFilter): PagedSnapshotInfos = sandboxService.listSnapshots(filter)

    fun deleteSnapshot(snapshotId: String) = sandboxService.deleteSnapshot(snapshotId)

    /**
     * Closes this resource, relinquishing any underlying resources.
     *
     * This method closes the local HTTP client resources associated with this sandbox manager instance.
     */
    override fun close() {
        try {
            httpClientProvider.close()
        } catch (e: Exception) {
            logger.warn("Error closing resources", e)
        }
    }

    class Builder internal constructor() {
        /**
         * Connection config
         */
        private var connectionConfig: ConnectionConfig? = null

        fun connectionConfig(connectionConfig: ConnectionConfig): Builder {
            this.connectionConfig = connectionConfig
            return this
        }

        /**
         * Creates the sandbox manager with the configured parameters.
         *
         * @return Fully configured SandboxManager instance
         * @throws InvalidArgumentException if required configuration is missing or invalid
         * @throws SandboxException if manager creation fails
         */
        fun build(): SandboxManager {
            return SandboxManager.create(
                connectionConfig = connectionConfig ?: ConnectionConfig.builder().build(),
            )
        }
    }
}
